packages/backend/src/sbom/sbom-fetching-utils.ts

```diff
@@ -0,0 +1,100 @@
+import PromisePool from "@supercharge/promise-pool";
+import axios from "axios";
+import axiosRetry from "axios-retry";
+import { logger } from "../utils/logger";
+import { Component, Dependency, PackageRequestData } from "./sbom-types";
+
+export const fetchPackages = async (
+  dependencies: Dependency[],
+  components: Component[]
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+): Promise<any[]> => {
+  // filter out deps without "dependsOn"
+  const withDependsOn = dependencies.filter((dep) => !!dep.dependsOn);
+
+  // map all dependsOn values and flat the matrix
+  const flatDeps = withDependsOn.map((dep) => dep.dependsOn).flat();
+
+  // filter out deps without "component"
+  const withComponent = flatDeps.filter((purl) =>
+    components.some((component) => component.purl === purl)
+  );
+
+  const npmPackages = withComponent
+    .filter((purl) => isNpmPackage(purl))
+    .map((purl) => resolvePackageRequestData(purl, components))
+    .filter((pkg) => !!pkg)
+    .map(fetchDataFromNPM);
+
+  const pyPackages = withComponent
+    .filter((purl) => isPythonPackage(purl))
+    .map((purl) => resolvePackageRequestData(purl, components))
+    .filter((pkg) => !!pkg)
+    .map(fetchDataFromPyPi);
+
+  const allPackagesPromises = [...npmPackages, ...pyPackages];
+  if (allPackagesPromises.length === 0) {
+    // no need to fetch anything
+    return [];
+  }
+
+  logger.sbom.log(`Fetching ${allPackagesPromises.length} packages...`);
+
+  // default concurrency is 10
+  const { results } = await PromisePool.for(allPackagesPromises).process(
```

Comment:
Isn't code in allPackagPromises already running at this point  ?
I think PromisePool.for should take objects, no existing Promises

Reply:
@itayox ? :)

Reply:
No, allPackagesPromises is just an array of promises. They are resolved in the `execute` callback with `await`.
Why? are you seeing something different?

Reply:
@itayox when you create a Promise, the code inside it is instantly executed, so if I'm not wrong (I'm only 80% sure ^^), all the GET are directly run, then the PromisePool.for just waits for results 10 by 10, but if there are 60 packages, the 60 axios GET already running before the PromisePool.for :/

Reply:
@itayox i checked with console.log & breakpoints and I confirm the Promise case is executed before the PromisePool.for :/
![image](https://github.com/oxsecurity/codetotal/assets/17500430/7a7a551c-9670-47fc-b97b-599032c2575f)

Reply:
@itayox i merged your branch in mine, I simplify all this code , i'll let you know here if I succeeded tonight :)